components/calendar/src/error.rs

```diff
@@ -21,12 +21,185 @@ pub enum DateError {
         /// The maximum value (inclusive). This might not be tight.
         max: i32,
     },
-    /// Unknown era
+    /// The era code is invalid for the calendar.
     #[displaydoc("Unknown era")]
     UnknownEra,
-    /// Unknown month code
+    /// The month code is invalid for the calendar or year.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_calendar::Date;
+    /// use icu_calendar::DateError;
+    /// use icu_calendar::cal::Hebrew;
+    /// use icu_calendar::types::MonthCode;
+    /// use tinystr::tinystr;
+    ///
+    /// let month_code = MonthCode(tinystr!(4, "M05L"));
+    ///
+    /// Date::try_new_from_codes(
+    ///     None,
+    ///     5784,
+    ///     month_code,
+    ///     1,
+    ///     Hebrew
+    /// )
+    /// .expect("5784 is a leap year");
+    ///
+    /// let err = Date::try_new_from_codes(
+    ///     None,
+    ///     5785,
+    ///     month_code,
+    ///     1,
+    ///     Hebrew
+    /// )
+    /// .expect_err("5785 is a common year");
+    ///
+    /// assert!(matches!(err, DateError::UnknownMonthCode(_)));
+    /// ```
     #[displaydoc("Unknown month code {0:?}")]
     UnknownMonthCode(MonthCode),
+    /// The year was specified in multiple inconsistent ways.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_calendar::Date;
+    /// use icu_calendar::DateError;
+    /// use icu_calendar::cal::Japanese;
+    /// use icu_calendar::types::DateFields;
+    /// use std::num::NonZeroU8;
+    ///
+    /// let mut fields = DateFields::default();
+    /// fields.era = Some("reiwa");
+    /// fields.era_year = Some(6);
+    /// fields.ordinal_month = NonZeroU8::new(1);
+    /// fields.day = NonZeroU8::new(1);
+    ///
+    /// Date::try_from_fields(
+    ///     fields,
+    ///     Default::default(),
+    ///     Japanese::new()
+    /// )
+    /// .expect("a well-defined Japanese date");
+    ///
+    /// fields.extended_year = Some(1900);
+    ///
+    /// let err = Date::try_from_fields(
+    ///     fields,
+    ///     Default::default(),
+    ///     Japanese::new()
+    /// )
+    /// .expect_err("year 1900 is not the same as 6 Reiwa");
+    ///
+    /// assert!(matches!(err, DateError::InconsistentYear));
+    /// ```
+    #[displaydoc("Inconsistent year")]
+    InconsistentYear,
```

Comment:
issue: these new variants are only returned by from-fields, not by constructors on `Date` (which is what this type is for). please lets not have catch-all errors again

Reply:
DateError is returned from Date::try_new_from_codes, which probably wouldn't exist if we had Date::try_from_fields first. So I'm basically changing DateError to do what Date::try_from_fields needs.

Reply:
If we kept from-codes and from-fields with different error types, what would you name them

Reply:
The current error stays `DateError`, `from_fields` returns `FromFieldsError`